File: static/py/components/base_component.py
# base_component.py
from pydantic import BaseModel, Field, PrivateAttr, validator
from typing import List, Optional, Union, Any
from abc import ABC, abstractmethod

from static.py.models.delete_me_later import ComponentProgressMessage, NavigationProgressMessage, OverallProgressMessage, SettingProgressMessage

def TrumpField(default: Any, trump: str):
    """Custom field with trump strategy for merge conflicts"""
    field = Field(default)
    field.field_info.extra["trump"] = trump
    return field

class BaseComponentConfig(BaseModel, ABC):
    # This class CANNOT be instantiated directly
    id: int = Field(..., ge=0, le=999999999999) #Immutable ID used to track components as they move for progress purposes
    type: str = Field(...) #Text, multiple choice question, task, etc.
    accessibility_label: str = Field(..., min_length=3) #Description for screen reader
    order: int = Field(...) #Components on a lesson go from least to greatest. Start with 100, 200, 300, .., to give room.

    @validator('type')
    @abstractmethod
    def validate_type_matches_class(cls, v):
        """Each child must implement type validation"""
        pass

class BaseComponentProgress(BaseModel, ABC):
    
    # Core lifecycle methods
    @abstractmethod
    def create_fields_for_config(self, config) -> dict:
        """Create all fields with proper initial values for this config"""
        pass

    # Merge strategy
    @abstractmethod
    def get_all_trump_strategies(self) -> dict:
        """Return trump strategy for every field in this component"""
        pass

class BaseComponentInternal(BaseModel, ABC):
    """Internal state that is never serialized or shared with core"""
    pass

class BaseComponent (ABC):
    def __init__(self, config: BaseComponentConfig, progress: BaseComponentProgress, timeline):
        self.config = config
        self.progress = progress
        self.timeline = timeline
        self.internal = self._create_internal_model()  # Must be implemented
        
        # Component creates its own slot and renders
        # TO DO add logic that interacts with timeline
        pass

    @abstractmethod
    def is_complete(self) -> bool:
        """Each component must implement this to let the navigator know if it is ready to move on"""
        pass

    @abstractmethod
    def _create_internal_model(self):
        """Each component must implement this to return its internal model instance"""
        pass
    
    def _render(self):
        # Build initial DOM in assigned container
        pass
        
    def get_component_progress_messages(self) -> List[ComponentProgressMessage]:
        # Core polling interface - return and clear message queue
        return []  # TODO: implement message queue

    def get_overall_progress_messages(self) -> List[OverallProgressMessage]:
        # Core polling interface - return and clear message queue
        return []  # TODO: implement message queue

    def get_navigation_messages(self) -> List[NavigationProgressMessage]:
        # Core polling interface - return and clear message queue
        return []  # TODO: implement message queue

    def get_setting_messages(self) -> List[SettingProgressMessage]:
        # Core polling interface - return and clear message queue
        return []  # TODO: implement message queue

    def destroy(self):
        if hasattr(self, 'timeline'):
            self.timeline.remove_component_slot(self.config.id)

File: src/ts/components/baseComponent.ts
// baseComponent.ts - TypeScript/Zod version of base_component.py
// Base classes for all learning components

import { z } from "zod";
import { TimelineContainer } from "../ui/timelineContainer.js";

/**
 * Trump strategy function type for merge conflicts
 */
export type TrumpStrategy<ProgressFieldType> = (
  a: ProgressFieldType,
  b: ProgressFieldType
) => ProgressFieldType;

/**
 * Base configuration schema that all components must extend
 * Cannot be instantiated directly - abstract equivalent
 * Field values derived from the lesson yaml for each component
 */

export const BaseComponentConfigSchema = z.object({
  id: z.number().int().min(0).max(999999999999), // Immutable ID for progress tracking
  type: z.string().min(1), // Component type: text, quiz, task, etc.
  accessibility_label: z.string().min(2), // Screen reader description
  order: z.number().int().min(-1000000).max(1000000), // Display order: 100, 200, 300, etc.
});

export type BaseComponentConfig = z.infer<typeof BaseComponentConfigSchema>;

/**
 * Base progress schema that all component progress must extend
 */

// No fields are universal for all components.
export const BaseComponentProgressSchema = z.object({});

// Could be any object with key value pairs or an empty object.
export type BaseComponentProgress = Record<string, any>;

/**
 * Base component progress class with methods
 * Abstract class that components must extend
 */
export abstract class BaseComponentProgressManager<
  TComponentProgress extends BaseComponentProgress
> {
  // Constructured from a complete version of itself
  constructor(protected progress: TComponentProgress) {}

  /**
   * Initially create all fields with proper initial values for this config. Eg: If the config has 3 checkboxes so does the schema.
   */
  abstract createInitialProgress(
    config: BaseComponentConfig
  ): TComponentProgress;

  /**
   * Get trump strategy for every field in this component
   */
  abstract getAllTrumpStrategies(): Record<
    keyof TComponentProgress,
    TrumpStrategy<any>
  >;
}

/**
 * Base internal state interface - never serialized or shared with core
 */
export interface BaseComponentInternal {
  rendered: boolean;
  [key: string]: any;
}

// Delete me and move later starting here =========================================================================================
// Placeholder message types (to be properly defined later)
export interface OverallProgressMessage {
  type: "overall_progress";
  data: any;
}

export interface ComponentProgressMessage {
  type: "overall_progress";
  data: any;
}

export interface NavigationProgressMessage {
  type: "navigation";
  data: any;
}

export interface SettingProgressMessage {
  type: "setting";
  data: any;
}
// End Delete me later here ========================================================================================================

/**
 * Abstract base component class
 * All interactive learning components must extend this
 */
export abstract class BaseComponent<
  TConfig extends BaseComponentConfig,
  TComponentProgress extends BaseComponentProgress,
  TInternal extends BaseComponentInternal
> {
  protected config: TConfig;
  protected progressManager: BaseComponentProgressManager<TComponentProgress>;
  protected internal: TInternal;
  protected timeline: TimelineContainer;

  constructor(
    config: TConfig,
    progressManager: BaseComponentProgressManager<TComponentProgress>,
    timeline: TimelineContainer
  ) {
    this.config = config;
    this.progressManager = progressManager;
    this.timeline = timeline;
    this.internal = this.createInternalModel();

    // Component creates its own slot and renders
    this.initializeComponent();
  }

  /**
   * Initialize the component - create slot and initial render
   */
  private initializeComponent(): void {
    // Add slot to timeline
    this.timeline.addComponentSlot(this.config.id.toString());

    // Initial render
    this.render();
  }

  /**
   * Each component must implement completion check
   */
  abstract isComplete(): boolean;

  /**
   * Each component must create its internal model
   */
  protected abstract createInternalModel(): TInternal;

  /**
   * Each component must implement rendering
   */
  protected abstract render(): void;

  /**
   * Get component progress messages for core polling
   */
  getComponentProgressMessages(): ComponentProgressMessage[] {
    // TODO: implement message queue
    return [];
  }

  /**
   * Get overall progress messages for core polling
   */
  getOverallProgressMessages(): OverallProgressMessage[] {
    // TODO: implement message queue
    return [];
  }

  /**
   * Get navigation messages for core polling
   */
  getNavigationMessages(): NavigationProgressMessage[] {
    // TODO: implement message queue
    return [];
  }

  /**
   * Get setting messages for core polling
   */
  getSettingMessages(): SettingProgressMessage[] {
    // TODO: implement message queue
    return [];
  }

  /**
   * Clean up component resources
   */
  destroy(): void {
    if (this.timeline) {
      this.timeline.removeComponentSlot(this.config.id.toString());
    }
  }

  /**
   * Get component area in timeline for custom rendering
   */
  protected getComponentArea(): HTMLElement | null {
    return this.timeline.getComponentArea(this.config.id.toString());
  }

  /**
   * Update component status in timeline
   */
  protected updateStatus(status: "active" | "completed" | "locked"): void {
    this.timeline.updateComponentStatus(this.config.id.toString(), status);
  }
}

/**
 * Component registry type for mapping types to classes
 */
export type ComponentRegistry = Map<
  string,
  new (...args: any[]) => BaseComponent<any, any, any>
>;

/**
 * Schema registry type for mapping types to Zod schemas
 */
export type SchemaRegistry = Map<
  string,
  {
    config: z.ZodSchema<any>;
    progress: z.ZodSchema<any>;
  }
>;

/**
 * Helper function to create trump field with strategy
 */
export function createTrumpField<T>(
  defaultValue: T,
  strategy: TrumpStrategy<T>
) {
  return {
    default: defaultValue,
    trump: strategy,
  };
}
